src/udev/dmi_memory_id/dmi_memory_id.c

```diff
@@ -0,0 +1,766 @@
+                if (h.type == 127)
+                        break;
+
+                /* Look for the next handle */
+                next = data + h.length;
+                while ((size_t)(next - buf + 1) < len && (next[0] != 0 || next[1] != 0))
+                        next++;
+                next += 2;
+
+                /* Make sure the whole structure fits in the table */
+                if ((size_t)(next - buf) > len) {
+                        data = next;
+                        break;
+                }
+
+                if (display)
+                        dmi_decode(&h);
+
+                data = next;
+
+                /* SMBIOS v3 requires stopping at this marker */
+                if (h.type == 127 && stop_at_eot)
+                        break;
+        }
+}
+
+static void dmi_table(int64_t base, uint32_t len, uint16_t num, const char *devmem,
+                bool stop_at_eot, bool no_file_offset) {
+        _cleanup_free_ uint8_t *buf = NULL;
+        size_t size;
+        int r;
+
+        /*
+         * When reading from sysfs or from a dump file, the file may be
+         * shorter than announced. For SMBIOS v3 this is expcted, as we
+         * only know the maximum table size, not the actual table size.
+         * For older implementations (and for SMBIOS v3 too), this
+         * would be the result of the kernel truncating the table on
+         * parse error.
+         */
+        r = read_full_file_full(AT_FDCWD, devmem,
+                        no_file_offset ? 0 : base, len,
+                        0, NULL, (char **) &buf, &size);
+        if (r < 0) {
+                log_error_errno(r, "Failed to read table, sorry: %m");
+                return;
+        }
+
+        len = size;
+        dmi_table_decode(buf, len, num, stop_at_eot);
+}
+
+/* Same thing for SMBIOS3 entry points */
+static int smbios3_decode(uint8_t *buf, const char *devmem, bool no_file_offset) {
+        uint64_t offset;
+
+        /* Don't let checksum run beyond the buffer */
+        if (buf[0x06] > 0x20) {
+                log_error("Entry point length too large (%u bytes, expected %u).",
+                          (unsigned)buf[0x06], 0x18U);
+                return 0;
+        }
+
+        if (!verify_checksum(buf, buf[0x06]))
+                return 0;
+
+        offset = QWORD(buf + 0x10);
+        if (!no_file_offset && (offset >> 32) != 0 && sizeof(int64_t) < 8) {
+                log_error("64-bit addresses not supported, sorry.");
+                return 0;
+        }
+
+        dmi_table(offset, DWORD(buf + 0x0C), 0, devmem, true, no_file_offset);
+
+        return 1;
+}
+
+static int smbios_decode(uint8_t *buf, const char *devmem, bool no_file_offset) {
+        /* Don't let checksum run beyond the buffer */
+        if (buf[0x05] > 0x20) {
+                log_error("Entry point length too large (%u bytes, expected %u).",
+                          (unsigned)buf[0x05], 0x1FU);
+                return 0;
+        }
```

Comment:
continuing from above, following our usual coding style we would code this as:

```c
if (buf[0x05] > 0x20) 
        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "…);
```

i.e. propagate an roughly reasonably error code (we wrap it in a SYNTHETIC_ERRNO() macro call in this case, which we do whenver we generate the error here, instead of propagating some error code we received from some other call which generated the error itself